commons/all/src/main/java/org/infinispan/commons/io/ExposedByteArrayOutputStream.java

```diff
@@ -1,37 +1,44 @@
 package org.infinispan.commons.io;
 
-import java.io.ByteArrayOutputStream;
+import java.io.OutputStream;
 
 import net.jcip.annotations.NotThreadSafe;
+import org.infinispan.commons.util.Util;
 
 /**
- * Extends ByteArrayOutputStream, but exposes the internal buffer. Using this, callers don't need to call toByteArray()
- * which copies the internal buffer. <p> Also overrides the superclass' behavior of always doubling the size of the
- * internal buffer any time more capacity is needed.  This class doubles the size until the internal buffer reaches a
- * configurable max size (default is 4MB), after which it begins growing the buffer in 25% increments.  This is intended
- * to help prevent an OutOfMemoryError during a resize of a large buffer. </p> <p> A version of this class was
- * originally created by Bela Ban as part of the JGroups library. </p> This class is not threadsafe as it will not
- * support concurrent readers and writers.
- * <p/>
+ * ByteArrayOutputStream alternative exposing the internal buffer. Using this, callers don't need to call toByteArray()
+ * which copies the internal buffer.
+ *
+ * <p> Also overrides the superclass' behavior of always doubling the size of the
+ * internal buffer any time more capacity is needed.
+ * This class doubles the size until the internal buffer reaches a
+ * configurable max size (default is 4MB), after which it begins growing the buffer in 25% increments.
+ * This is intended to help prevent an OutOfMemoryError during a resize of a large buffer. </p>
+ * <p> A version of this class was originally created by Bela Ban as part of the JGroups library. </p>
+ * <p>This class is not threadsafe as it will not support concurrent readers and writers. <p/>
  *
  * @author <a href="mailto://brian.stansberry@jboss.com">Brian Stansberry</a>
  * @since 4.0
  */
 @NotThreadSafe
-public final class ExposedByteArrayOutputStream extends ByteArrayOutputStream {
+public final class ExposedByteArrayOutputStream extends OutputStream {
```

Comment:
this is public API. I don't think we can make this change.
https://docs.jboss.org/infinispan/12.1/apidocs/org/infinispan/commons/io/package-summary.html

Reply:
You're right... I was thought I could get away with it because `ByteArrayOutputStream` just implements `OutputStream`, but it does add some public methods, which means users may declare variables of type `ByteArrayOutputStream`.

I can't delay the allocation if I keep extending `ByteArrayOutputStream`, so I think I'll have to add a new class.

Reply:
the `byte[] buf` from `ByteArrayOutputStream` is available directly. Can't you control how it grows?

Reply:
Copied to `LazyByteArrayOutputStream`